Upgrade to Airflow 2.0 on Astronomer Doc (#130)
astronomer/docs: cloud/stable/04_customize-airflow/01_upgrade-to-airflow-2.md

- Simplified KubernetesExecutor for ultimate flexibility in configuration.
- [UI/UX Improvements](https://github.com/apache/airflow/pull/11195) including a new Airflow UI and auto-refresh button in the **Graph** view.

### Airflow 2.0 Resources

For more information on Airflow 2.0, we highly recommend the following:

- ["Introducing Airflow 2.0" blog post](https://www.astronomer.io/blog/introducing-airflow-2-0) from Astronomer
- [Airflow 2.0 YouTube Series](https://www.youtube.com/playlist?list=PLCi-q9vYo4x-PESoBcXN0tXCMgzh5c_Pj) from Astronomer
- Airflow 2.0 blog post from the Apache Airflow Project

As always, don't hesitate to reach out to [Astronomer Support](https://support.astronomer.io) for questions.

## Step 1: Upgrade to Airflow 1.10.14

Airflow 1.10.14, considered a "bridge" release to Airflow 2.0, was built to make the migration and testing process as easy as possible. On Astronomer, you MUST upgrade to an Astronomer Certified (AC) 1.10.14 image before upgrading to any 2.0 version.

To upgrade to 1.10.14,

1. Initialize the Airflow upgrade process via the Astronomer UI or CLI
2. Deploy the AC 1.10.14 image (Debian-only):

> ```

Reviewer: ```suggestion
```dockerfile
```

Now that we have syntax highlighting :)

Author: Nice call - added!

### a. Run the Upgrade Check from the Astronomer CLI

To run the Airflow 2.0 upgrade check script, run:

```
$ astro dev upgrade-check
```

This command will automatically install the latest version of the `apache-airflow-upgrade-check` package at runtime and output results from the corresponding Airflow CLI command. It is only supported locally.

#### b. Review Upgrade Check Recommendations

In response to the command above, you should see something like the following:

```
==============================================STATUS============================================

Check for latest versions of apache-airflow and checker..................................SUCCESS
Remove airflow.AirflowMacroPlugin class..................................................SUCCESS
Chain between DAG and operator not allowed...............................................SUCCESS
Connection.conn_id is not unique.........................................................SUCCESS
Connection.conn_type is not nullable.....................................................SUCCESS

> Fernet is enabled by default.............................................................FAIL

Reviewer 2: Will this ever fail in an Astronomer deployment? 
I was under the impression that we always specify a Fernet key.

GCP service account key deprecation......................................................SUCCESS
Changes in import paths of hooks, operators, sensors and others..........................FAIL

> Users must delete deprecated configs for KubernetesExecutor..............................FAIL

Reviewer 2: I think it is worth validating how this would be different in an Astronomer deployment. 
Which of these settings are actively used?

Legacy UI is deprecated by default.......................................................SUCCESS
Logging configuration has been moved to new section......................................FAIL
Removal of Mesos Executor................................................................SUCCESS
Users must set a kubernetes.pod_template_file value......................................FAIL
SendGrid email uses old airflow.contrib module...........................................SUCCESS
Changes in import path of remote task handlers...........................................SUCCESS
Jinja Template Variables cannot be undefined.............................................FAIL
Found 7 problems.

========================================== RECOMMENDATIONS ========================================

Fernet is enabled by default
----------------------------
The fernet mechanism is enabled by default to increase the security of the default installation.

Problems:

> 1.  fernet_key in airflow.cfg must be explicitly set empty as fernet mechanism is enabled by default.

Reviewer 2: Related to the question above. Is this even controllable by the user directly through the airflow.cfg file? 

If it not a problem within Astronomer I think we can delete this section from this document.

This means that the apache-airflow[crypto] extra-packages are always installed.However, this requires
that your operating system has libffi-dev installed.

Changes in import paths of hooks, operators, sensors and others
---------------------------------------------------------------
Many hooks, operators and other classes has been renamed and moved. Those changes were part of
unifying names and imports paths as described in AIP-21.
The contrib folder has been replaced by providers directory and packages:
https://github.com/apache/airflow#backport-packages

Problems:

1.  Using ``airflow.operators.python_operator.PythonOperator`` will be replaced by ``airflow.operators.python.PythonOperator``. Affected file:
```

For more information on upgrade check functionality, read [Upgrade Check Script](
]https://airflow.apache.org/docs/apache-airflow/stable/upgrade-check.html#upgrade-check) from Apache Airflow.

> 

Reviewer 2: I think the one step missing here is the changes which otherwise would be done to the airflow.cfg file in an Astronomer environment. Effectively, my question is around the moving of the settings regarding logging configuration (possibly irrelevant) and the kubernetes settings (pod_template_file).

## Step 3: Prepare Airflow 2.0 DAGs

The next step is to review the results from Airflow's upgrade check script and make all necessary changes to your import statements, DAGs, and configurations.

a. Import Operators from Backport Providers

All Airflow 2.0 Operators are backwards compatible with Airflow 1.10. To transition to using these providers:

> 1. Add the new providers to your `requirements.txt` or `packages.txt` files

Reviewer 2: Assuming people running this are using our 1.10.14 AC docker image, does this reduce the burden in any way because of the included providers? 
I am thinking out loud here - a question not a statement.

---
title: "Upgrade to Apache Airflow 2.0 on Astronomer"
navTitle: "Upgrade to Airflow 2.0"
description: "How to prepare for and upgrade to Airflow 2.0 on Astronomer."
---

## Overview

Apache Airflow 2.0 is a momentous open-source release that we're thrilled to support. For those eager to learn more about Airflow 2.0 and upgrade an existing Airflow Deployment on Astronomer, read the guidelines below.

At a high-level, the recommended upgrade path is as follows:

1. Upgrade to Python 3

> 2. Migrate to Airflow 1.10.14

Reviewer: Should we say AC 1.10.14 here? Or does AC / Airflow not matter here?

Author: I don't think it matters here, but I do talk about AC below!

---

> title: "Upgrade to Airflow 2.0 on Astronomer"

Reviewer 3: ```suggestion
title: "Upgrade to Apache Airflow 2.0 on Astronomer"
```

navTitle: "Upgrade to Airflow 2.0"
description: "How to prepare for and upgrade to Airflow 2.0 on Astronomer."
---

## Overview

> Airflow 2.0 is a momentous open-source release that we're thrilled to support. For those eager to learn more about Airflow 2.0 and upgrade an existing Airflow Deployment on Astronomer, read the guidelines below.

Reviewer 3: ```suggestion
Apache Airflow 2.0 is a momentous open-source release that we're thrilled to support. For those eager to learn more about Airflow 2.0 and upgrade an existing Airflow Deployment on Astronomer, read the guidelines below.
```

After these first two references it's probably okay to just call it "Airflow"

At a high-level, the recommended upgrade path is as follows:

> 1. Upgrade to Python 3

Reviewer 3: Our users are already on py3 -- we've never published a py2 image I don't think.

Author: Oh damn, good call - oversight on my part. I'll remove this section.

2. Migrate to Airflow 1.10.14
3. Run the Airflow Upgrade Check Script
4. Modify your DAGs, Configs, and Import Statements
5. Upgrade to Airflow 2.0

Following a section that highlights the biggest features included in Airflow 2.0, this doc will walk through each step listed above.

> **Note:** The recommendations around the upgrade process are largely written in collaboration with the Apache Airflow project and community, though there a few differences in steps on Astronomer that we've made sure to outline below.

## Why Airflow 2.0

Airflow 2.0 is built to be faster, more reliable, and significantly more scalable than 1.10. Among the hundreds of new features both large and small, Airflow 2.0 includes:

> - [Refactored Airflow Scheduler](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/scheduler.html#running-more-than-one-scheduler) for enhanced performance and high-availability.
> - [Full REST API](https://airflow.readthedocs.io/en/latest/stable-rest-api-ref.html) that enables opportunity for automation.
> - [Smart Sensors](https://airflow.readthedocs.io/en/latest/smart-sensor.html) that execute as single, long-running tasks.
> - [TaskFlow API](https://airflow.readthedocs.io/en/latest/concepts.html#taskflow-api) for a simpler way to pass information between tasks.

Reviewer 3: All these links will need updating.

```suggestion
- [Refactored Airflow Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/scheduler.html#running-more-than-one-scheduler) for enhanced performance and high-availability.
- [Full REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html) that enables opportunity for automation.
- [Smart Sensors](https://airflow.apache.org/docs/apache-airflow/stable/smart-sensor.html) that execute as single, long-running tasks.
- [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskflow-api) for a simpler way to pass information between tasks.
```

Those won't work yet though

- [Independent Providers](https://github.com/apache/airflow/tree/master/airflow/providers) for improved usability and a more agile release cadence.
- Simplified KubernetesExecutor for ultimate flexibility in configuration.
- [UI/UX Improvements](https://github.com/apache/airflow/pull/11195) including a new Airflow UI and auto-refresh button in the **Graph** view.

### Airflow 2.0 Resources

For more information on Airflow 2.0, we highly recommend the following:

> - ["Introducing Airflow 2.0"](https://www.astronomer.io/blog/introducing-airflow-2-0) Blog Post from Astronomer

Reviewer 3: ```suggestion
- ["Introducing Airflow 2.0" blog post](https://www.astronomer.io/blog/introducing-airflow-2-0) from Astronomer
```

- [Airflow 2.0 YouTube Series](https://www.youtube.com/playlist?list=PLCi-q9vYo4x-PESoBcXN0tXCMgzh5c_Pj) from Astronomer

> - Official Airflow 2.0 Changelog

Reviewer 3: This should be the release announcement once I publish it. (Not sure there will be a changelog in a timely fashion for 2.0

Author: Yep I was planning on it Ash! I'll change language from "Changelog" - thanks

For guidance on upgrading Python, here are some recommended resources:

- ["Migrating Slack Airflow to Python 3 Without Disruption"](https://slack.engineering/migrating-slack-airflow-to-python-3-without-disruption/) from Slack Engineering
- ["Tips and Tricks for Upgrading from Python 2 to Python 3"](https://blog.couchbase.com/tips-and-tricks-for-upgrading-from-python-2-to-python-3/) from the CouchBaseDB team

> **Note:** If you have a specific task that still requires Python 2, you can use the [PythonVirtualenvOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/python_operator/index.html) or the [KubernetesPodOperator](https://www.astronomer.io/docs/cloud/stable/customize-airflow/kubepodoperator).

## Step 2: Upgrade to Airflow 1.10.14

Airflow 1.10.14, considered a "bridge" release to Airflow 2.0, was built to make the migration and testing process as easy as possible. On Astronomer, you MUST upgrade to an Astronomer Certified (AC) 1.10.14 image before upgrading to any 2.0 version.

To upgrade to 1.10.14,

1. Initialize the Airflow upgrade process via the Astronomer UI or CLI

> 2. Deploy the AC 1.10.14 image (Debian-only):

Reviewer 3: We probably need to expand what we mean by "debian only" here.

Author: Yea, I can add a note here.

```
FROM quay.io/astronomer/ap-airflow:1.10.14-buster-onbuild
```

For detailed guidelines on how to upgrade Airflow on Astronomer, read [“Manage Airflow Versions”](https://www.astronomer.io/docs/cloud/stable/customize-airflow/manage-airflow-versions). For more information on 1.10.14, check out the [Airflow Release](https://github.com/apache/airflow/releases/tag/1.10.14) or the corresponding [AC 1.10.14 changelog](https://github.com/astronomer/ap-airflow/blob/master/1.10.14/CHANGELOG.md).

## Step 3: Run the Airflow Upgrade Check Script

> Once you've upgraded to 1.10.14, you're ready to start making your DAGs and configurations Airflow 2.0-compatible. Most Airflow 2.0-compatible DAGs will work in Airflow 1.10.14, though not all.

Reviewer 3: This should be `astro dev upgrade_check` or something for us.

Author: Yep - specify that below!

> For a comprehensive list of recommended and required changes specific to your existing project, run the Airflow upgrade check script via the Astronomer CLI. The script reviews your DAG code, deployment-level configuration from `airflow.cfg`, and metadata from the Airflow Database.

Reviewer 3: This needs tweaking -- our users can't set airflow.cfg (it's over-ridden) so they'll... I don't  know what. Have to check something about the env vars they set in their deployments.

Author: Yea, I thought about that being misleading here. Changed to this:

> For a comprehensive list of recommended and required changes specific to your existing project, run the Airflow upgrade check script via the Astronomer CLI. The script reviews your DAG code, deployment-level configuration and Environment Variables, as well as metadata from the Airflow Database.

Let me know if that's any better. It sounds like it'll be likely for Astro users to see airflow.cfg/config changes in the output of this upgrade check - maybe we should clarify how people should handle those? I presume some of them will be not applicable as well?
